api/v1alpha1/author_webhook.go

```diff
@@ -0,0 +1,197 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+	"strings"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	"github.com/redhat-appstudio/release-service/metadata"
+	admissionv1 "k8s.io/api/admission/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+const MAX_LABEL_LENGTH = 63
+
+// authorWebhook describes the data structure for the author webhook
+type authorWebhook struct {
+	Client client.Client
+	log    *logr.Logger
+}
+
+// +kubebuilder:webhook:path=/mutate-appstudio-redhat-com-v1alpha1-author,mutating=true,failurePolicy=fail,sideEffects=None,groups=appstudio.redhat.com,resources=releases;releaseplans,verbs=create;update,versions=v1alpha1,name=mauthor.kb.io,admissionReviewVersions=v1
+
+// RegisterAuthorWebhook registers the author webhook with the passed manager and log.
+func RegisterAuthorWebhook(mgr ctrl.Manager, log logr.Logger) {
+	mgr.GetWebhookServer().Register("/mutate-appstudio-redhat-com-v1alpha1-author", &webhook.Admission{
+		Handler: &authorWebhook{
+			Client: mgr.GetClient(),
+			log:    &log,
+		},
+	})
+}
+
+// Handle creates an admission response for Release and ReleasePlan requests.
+func (a *authorWebhook) Handle(ctx context.Context, req admission.Request) admission.Response {
+	switch req.Kind.Kind {
+	case "Release":
+		return a.handleRelease(req)
+	case "ReleasePlan":
+		return a.handleReleasePlan(req)
+	default:
+		return admission.Errored(http.StatusInternalServerError, errors.New("webhook tried to handle an unsupported resource: "+string(req.Kind.Kind)))
+	}
+}
+
+// handleRelease takes an incoming admission request and returns an admission response. Create requests
+// add an author label with the current user. Update requests are rejected if the spec or author label
+// are being modified. All other requests are accepted without action.
+func (a *authorWebhook) handleRelease(req admission.Request) admission.Response {
+	var releaseNew Release
+	err := json.Unmarshal(req.Object.Raw, &releaseNew)
+	if err != nil {
+		return admission.Errored(http.StatusInternalServerError, errors.Wrap(err, "error decoding object"))
+	}
+	switch req.AdmissionRequest.Operation {
+	case admissionv1.Create:
+		return a.patchAuthorResponse(req.UserInfo.Username, &releaseNew)
+	case admissionv1.Update:
+		var releaseOld Release
+		err = json.Unmarshal(req.OldObject.Raw, &releaseOld)
```

Comment:
You don't need this, isn't it? You already have it at the beginning of this function.

Reply:
One is the object, one is the oldobject